modules/codegen/src/smithy4s/codegen/SmithyToIR.scala

```diff
@@ -253,23 +254,59 @@ private[codegen] class SmithyToIR(model: Model, namespace: String) {
           shapeId.getName()
         )
 
+      private def getExternalTypeInfo(
+          shape: Shape
+      ): Option[RefinedTrait] = {
+        shape
+          .getAllTraits()
+          .asScala
+          .flatMap { case (_, trt) =>
+            model
+              .getShape(trt.toShapeId)
+              .asScala
+              .flatMap(_.getTrait(classOf[RefinedTrait]).asScala.flatMap {
+                refTrait =>
+                  if (
+                    refTrait.getCanonicalShapeId.asScala.contains(shape.getId)
+                  ) // Only return if is canonical shape
+                    Some(refTrait)
+                  else None
+              })
+          }
+          .headOption // TODO: Currently shapes could have multiple traits that have the refined trait. A validator should be added to prevent this
+      }
+
       def primitive(
           shape: Shape,
           primitiveId: String,
           primitive: Primitive
       ): Option[Type] = {
-        if (
-          shape.getId() != ShapeId
-            .from(primitiveId) && !isUnboxedPrimitive(shape.getId())
-        ) {
-          Type
-            .Alias(
-              shape.getId().getNamespace(),
-              shape.getId().getName(),
-              Type.PrimitiveType(primitive)
+        val res =
+          if (
+            shape.getId() != ShapeId
+              .from(primitiveId) && !isUnboxedPrimitive(shape.getId())
+          ) {
+            Type
+              .Alias(
+                shape.getId().getNamespace(),
+                shape.getId().getName(),
+                Type.PrimitiveType(primitive)
+              )
+              .some
+          } else Type.PrimitiveType(primitive).some
+
+        getExternalTypeInfo(shape) match {
```

Comment:
we may want to allow it on collection types too, to enable things like `NonEmptyList` and whatnot.